cache/network/api.go

```diff
@@ -0,0 +1,230 @@
+
+	body, err := json.Marshal(requestBody)
+	if err != nil {
+		return prepareUploadResponse{}, err
+	}
+
+	req, err := retryablehttp.NewRequest("POST", url, body)
+	if err != nil {
+		return prepareUploadResponse{}, err
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.accessToken))
+	req.Header.Set("Content-type", "application/json")
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return prepareUploadResponse{}, err
+	}
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(resp.Body)
+
+	if resp.StatusCode != http.StatusCreated {
+		errorResp, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return prepareUploadResponse{}, err
+		}
+		return prepareUploadResponse{}, fmt.Errorf("HTTP %d: %s", resp.StatusCode, errorResp)
+	}
+
+	var response prepareUploadResponse
+	err = json.NewDecoder(resp.Body).Decode(&response)
+	if err != nil {
+		return prepareUploadResponse{}, err
+	}
+
+	return response, nil
+}
+
+func (c apiClient) uploadArchive(archivePath, uploadMethod, uploadURL string, headers map[string]string) error {
+	file, err := os.Open(archivePath)
+	if err != nil {
+		return err
+	}
+
+	req, err := retryablehttp.NewRequest(uploadMethod, uploadURL, file)
+	if err != nil {
+		return err
+	}
+	for k, v := range headers {
+		req.Header.Set(k, v)
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		errorResp, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, errorResp)
+	}
+	return nil
+}
+
+func (c apiClient) acknowledgeUpload(uploadID string) error {
+	url := fmt.Sprintf("%s/upload/%s/acknowledge", c.baseURL, uploadID)
+
+	req, err := retryablehttp.NewRequest("PATCH", url, nil)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.accessToken))
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		errorResp, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, errorResp)
+	}
+	return nil
+}
+
+func (c apiClient) restore(cacheKeys []string) (string, error) {
+	keysInQuery, err := validateKeys(cacheKeys)
+	if err != nil {
+		return "", err
+	}
+	apiURL := fmt.Sprintf("%s/restore?cache_keys=%s", c.baseURL, keysInQuery)
+
+	req, err := retryablehttp.NewRequest("GET", apiURL, nil)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.accessToken))
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(resp.Body)
+
+	if resp.StatusCode == http.StatusNotFound {
+		return "", ErrCacheNotFound
+	}
+	if resp.StatusCode != http.StatusOK {
+		errorResp, err := ioutil.ReadAll(resp.Body)
```

Comment:
This block is repeated many times, what about creating a function for it?

Reply:
Check out the latest commit, I created a function for this